client/src/Components/Notebook.jsx

```diff
@@ -4,28 +4,28 @@ import Container from "react-bootstrap/Container";
 import NavigationBar from "./Shared/NavigationBar";
 import uuidv4 from "uuid";
 
-import {
-  findSyntaxErrorIdx,
-  findPendingIndex,
-  findWriteToPendingIndex,
-  findCellIndex,
-  findLastIndexOfEachLanguageInNotebook,
-  makeStopPendingObj
-} from "../utils";
+import { findLastIndexOfEachLanguageInNotebook } from "../utils";
 import { SIGTERM_ERROR_MESSAGE, PROXY_URL } from "../Constants/constants";
 
 class Notebook extends Component {
   state = {
     cells: [],
-    RubyPendingIndexes: [],
-    RubyWriteToIndex: 0,
-    RubyCodePending: false,
-    JavascriptPendingIndexes: [],
-    JavascriptWriteToIndex: 0,
-    JavascriptCodePending: false,
-    PythonPendingIndexes: [],
-    PythonWriteToIndex: 0,
-    PythonCodePending: false,
+    Ruby: {
```

Comment:
This is great! Cool how a minor change in the shape of state can make a bunch of helper functions unnecessary